### DISCUSSION TITLE
Issues with Many parser when reading from file.

### DISCUSSION THREAD

**Discussion by lukeredpath** [pointfreeco/swift-parsing#260] [Q&A]
I'm running into an issue when running a parser against input from a file. I'm using swift-parsing for this years Advent Of Code, as I did last year, but last year I configured my input in swift files using a string. This year I'm reading them from plain text files and every single parser I write for what should be a trivial "parse many of this on each line" kind of parser fails when reading from a file, but not with sample data.

For example given this example input:

```swift
let sampleInput = """
    2-4,6-8
    2-3,4-5
    5-7,7-9
    2-8,3-7
    6-6,4-6
    2-6,4-8
    """
```

I can successfully parse this with:

```swift
typealias Assignment = Range<Int>

enum Parsers {
    static let assignment = Parse {
        Int.parser()
        "-"
        Int.parser()
    }.map(Assignment.init(uncheckedBounds:))
    
    static let assignmentPair = Parse {
        assignment
        ","
        assignment
    }
    
    static let input = Many {
        assignmentPair
    } separator: {
        Whitespace(1, .vertical)
    }
}
```

However, when I run this on my input file, it fails with:

```
Unexpected error: error: unexpected input
    --> input:1000:12
1000 | 29-31,12-28
     |            ^ expected end of input
```

My only workaround so far is to wrap the many parser like this, but it feels like it shouldn't be necessary:

```swift
Many {
    someParser
} separator: {
    Whitespace(1, .vertical)
}
Whitespace(1, .vertical)
End()
```

**Comment by lukeredpath** [comment]
It looks like an extra new line was getting inserted into the file - I solved this by trimming the string of new lines and whitespace when reading it from the file, before I pass it into the parser. Pretty weird though.

**Reply by stephencelis** [reply]
@lukeredpath I think it's quite common for files like this to contain a trailing newline.

I think what you're looking for is the `terminator` argument of `Many`, though:

```swift
Many {
    someParser
} separator: {
    Whitespace(1, .vertical)
} terminator: {
    Whitespace(1, .vertical)
}
```

As a bonus, specifying `terminator` helps produce better, more localized errors.
